internal/am/poll_transfer.go

```diff
@@ -0,0 +1,115 @@
+package am
+
+import (
+	context "context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	"go.artefactual.dev/amclient"
+	temporal_tools "go.artefactual.dev/tools/temporal"
+	temporalsdk_activity "go.temporal.io/sdk/activity"
+)
+
+const PollTransferActivityName = "poll-transfer-activity"
+
+type PollTransferActivityParams struct {
+	TransferID string
+}
+
+type PollTransferActivity struct {
+	logger logr.Logger
+	cfg    *Config
+	amts   amclient.TransferService
+}
+
+type PollTransferActivityResult struct {
+	SIPID string
+	Path  string
+}
+
+func NewPollTransferActivity(logger logr.Logger, cfg *Config, amts amclient.TransferService) *PollTransferActivity {
+	return &PollTransferActivity{logger: logger, cfg: cfg, amts: amts}
+}
+
+// Execute polls Archivematica for the status of a transfer and returns when
+// the transfer is complete or returns an error status. Execute sends an
+// activity heartbeat after each poll.
+//
+// A transfer status of "REJECTED", "FAILED", "USER_INPUT", or "BACKLOG" return
+// a temporal.NonRetryableApplicationError to indicate that processing can not
+// continue.
+func (a *PollTransferActivity) Execute(ctx context.Context, params *PollTransferActivityParams) (*PollTransferActivityResult, error) {
+	a.logger.V(1).Info("Executing PollTransferActivity", "TransferID", params.TransferID)
+
+	for {
+		resp, httpResp, err := a.amts.Status(ctx, params.TransferID)
+		if ferr := transferFailedError(httpResp, err); ferr != nil {
+			a.logger.V(2).Info("Poll transfer error",
+				"StatusCode", httpResp.StatusCode,
+				"Status", httpResp.Status,
+			)
+			return nil, ferr
+		}
+
+		complete, err := isComplete(resp)
+		if err != nil {
+			return nil, err
+		}
+		if complete {
+			return &PollTransferActivityResult{SIPID: resp.SIPID, Path: resp.Path}, nil
+		}
+
+		// Send a heartbeat then continue polling after the poll interval.
+		temporalsdk_activity.RecordHeartbeat(ctx,
+			fmt.Sprintf("PollTransferActivity: last HTTP response: %s", httpResp.Status),
+		)
+		time.Sleep(a.cfg.PollInterval)
+	}
+}
```

Comment:
A problem with `time.Sleep` is that it blocks and cannot be cancelled via context. It's not a huge deal but if you wanted to react to cancellation immediately then I'd suggest a different approach that is quite common where you use a `select` in combination with a timer, while you're also receiving from the context channel. I haven't tested but I think you'll get the idea. I'm also suggesting to pass a new context with a timeout to amclient which can be important to make sure that you're never missing a heartbeat deadline.

```go
func (a *PollTransferActivity) Execute(ctx context.Context, params *PollTransferActivityParams) (*PollTransferActivityResult, error) {
	ticker := time.NewTicker(a.cfg.PollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-ticker.C:
			reqCtx, cancel := context.WithTimeout(ctx, a.cfg.PollInterval/2)
			r, err := a.poll(reqCtx, params.TransferID, cancel)
			if err == errWorkOngoing {
				// Send a heartbeat then continue polling after the poll interval.
				temporalsdk_activity.RecordHeartbeat(ctx, fmt.Sprintf("Last HTTP response: %v", errWorkOngoing))
				continue
			}
			if err != nil {
				return nil, err
			}

			return r, nil
		}
	}
}
```

I chose to pass the cancel function to a.poll so you can use defer. Otherwise you'd need to call it from the three exit path: ongoing, error and happy path.

Reply:
Does it block even if the activity itself is cancelled?

Reply:
I believe so. Temporal can only signal cancellation via ctx (only true for heartbeating activities). It can't interrupt the execution without collaboration from the activity. 

That's not necessarily true in every SDK, but I suspect the root cause is that the Go SDK uses goroutines which need some coordination with the goroutine itself to be interrupted, e.g. a channel, context, etc... Temporal adopted context because it's the adopted idiom in the community.

David's activity eventually stops hence is not a leak, but it's making use of a goroutine unnecessarily for a period of time.

Related: https://dave.cheney.net/2016/12/22/never-start-a-goroutine-without-knowing-how-it-will-stop.

Reply:
Thanks. Something to test later as well.